ENH: added Moyal distribution to scipy.stats

seanpquinn opened the pull request:
A useful distribution for high-energy physics,
but also applied in a wide range of phenomena.

josef-pkt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4086,6 +4086,69 @@ def _stats(self, a):
+    def _pdf(self, x, b):
+        #moyal.pdf(x, b) = 1 / (sqrt(2 * pi) * b) * exp(-0.5 * (x / b
+        #+ exp(-x / b)))
+        earg = x / b
```
b is the scale parameter and is automatically added by the super class and needs to be dropped here

josef-pkt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4086,6 +4086,69 @@ def _stats(self, a):
+
+    def _cdf(self, x, b):
+        #moyal.cdf(x, b) = 1 - gammainc(0.5, 0.5 * exp(-x / b))
+        return 1 - sc.gammainc(0.5, 0.5 * np.exp(-x / b))
```
If there is no scipy.special function that can be used for ppf, then an explicit rvs method would be much faster.
http://www.stat.rice.edu/~dobelman/textfiles/DistributionsHandbook.pdf  (One of the first links I found with Google)
describe a rejection method based on two uniform random variables. Those kind of methods would fit better in cython as in numpy.random

josef-pkt commented:
extra info: 
http://threeplusone.com/FieldGuide.pdf   has it as a special case of the Gamma-exponential distribution section 7.9

josef-pkt commented:
http://reference.wolfram.com/language/ref/MoyalDistribution.html
lists it as a linear-log transformation of the gamma distribution 
z = loc + scale * log(u), 
where u ~ GammaDistribution[1/2, 2], 

I guess this should also work for creating random variables.

josef-pkt commented:
test failure with complex input, There is somewhere a black list to avoid checking complex in the unit tests, unless gammainc is replaced by erf, as in the `sf` which AFAIR works for complex

```
>       return 1 - sc.gammainc(0.5, 0.5 * np.exp(-x / b))
E       TypeError: ufunc 'gammainc' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```
x is complex

overall: adding the distribution sounds fine to me

needs more work, possibly adding rvs and ppf.
My guess is that the log transformation of a Gamma distributed variable can be used to get additional methods.

rkern commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4086,6 +4086,69 @@ def _stats(self, a):
 kappa3 = kappa3_gen(a=0.0, name='kappa3')
 
+class moyal_gen(rv_continuous):
+    r"""A Moyal continous random variable.
```
`continuous`

rkern commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4086,6 +4086,69 @@ def _stats(self, a):
+        return coef * np.exp(-0.5 * (earg + np.exp(-earg)))
+
+    def _cdf(self, x, b):
+        #moyal.cdf(x, b) = 1 - gammainc(0.5, 0.5 * exp(-x / b))
```
This comment (and the similar one in `_pdf`) just recapitulates the code that follows. You can delete them.

seanpquinn commented:
Thanks for all the wonderful suggestions. I hope this update addresses the issues:

1. Added rvs override based on @josef-pkt suggestion
2. Replaced `sc.gammainc` with `sc.erfc` in the CDF. Fixes complex unit test failure
3. Fixed typos and cleaned up comments, thanks @rkern

josef-pkt commented:
The change looks good to me. All main methods have now explicit computation. 

There might be some other methods that could be checked. For example, non-central moments should be available in closed form, at least for the first few.

seanpquinn commented:
A quick update:

1. Added first 4 raw moments, borrowed `n>4` from `truncexpon_gen`
2. Skew and kurt were wrong in initial commit I think. Fixed here.
3. Added additional reference to distribution guide
4. Fixed norm constant in notes

josef-pkt commented:
LGTM, I think it's finished from what I can think of on the statistics side, assuming unit tests verify the computations (I didn't check any formulas).

seanpquinn commented:
Ok, I think I am happy with things now. Tests should be OK. Sorry about wasting a non-trivial number of clock cycles getting it correct :disappointed:

chrisb83 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4086,6 +4086,89 @@ def _stats(self, a):
+        g2 = 4.
+        return mu, mu2, g1, g2
+
+    def _munp(self, n):
```
Hi, I might be wrong but it look like _munp returns the _central_ moments if n=2,3,4 whereas stats.moments should return _non-central_ moments (see 26.4 in http://www.stat.rice.edu/~dobelman/textfiles/DistributionsHandbook.pdf).

Christoph

seanpquinn replied:
Hi Christoph,

Thanks for your comments.

Hmm, oops, perhaps I made a mistake in `_munp`. For these I'm assuming the scaling/shifting is handled automatically based on the central moments. Otherwise it would be necessary to explicitly multiply the `_munp` parameters by appropriate powers of the scale parameter, etc. Should `_stats` really use raw moments? Perhaps an expert can weigh in here.

josef-pkt replied:
stats is fine and should return mvsk, mean, variance (centered second moment) skew and kurtosis

munp (if that is the right function name) should return non-centered moments. Note centering on mean is independent of loc scale in asymmetric distributions like this, super class takes care of loc and scale.
But in asymmetric distribution, mean is not zero, even if loc =0.

What is strange is that a bug here should have been caught by the generic unit tests. I haven't looked much at the details in recent time, but this is a part that can be verified generically.

chrisb83 replied:
Hi,

yes, it is not about `loc` or `scale`. munp should return E[X^k] and not E[(X - mean(X))^k] (central moment). In the handbook referenced above, the former is called mu_k' and the latter simply mu_k (section 2.2) and section 26.4 on the moyal distribution gives an expression for mu_2 and mu_3 (i.e. the centered one). (You can derive the non-centered moment by expanding the square / cube)

seanpquinn replied:
Ok, I have attempted to put in the correct analytic raw moments. With these changes if I look at a test distribution I get the following:

```
In [1]: import scipy.stats as ss
In [2]: mo=ss.moyal(loc=2,scale=2)
In [3]: for i in range(2,5):
    print "%.16f" %mo.moment(i)
    print "%.16f" %mo.expect(lambda x: x**i)
    print "  "
   ....:     
40.3573986023864748
40.3573986023866880
  
497.1429148596399727
497.1429148596396317
  
8039.7631977583432672
8039.7631977583487242
```

chrisb83 replied:
OK, Looks good. I don't know if the rather long expressions in case `n==3` or `n == 4` add much value since the expectation can be computed also numerically and the results are the same up to ~9 digits. `n==3`: note that you sometimes did not put spaces around the operator `+`. 
But apart from such very minor things, I think the implementation looks good.

seanpquinn replied:
In an earlier comment Josef mentioned first few raw moments should be available in closed form. I guess my definition of "few" was first four moments.

The expressions are long, but the speed-up should be significant compared to integration. I will fix the formatting around the operators, good catch.

josef-pkt commented:
When checking unit tests, I often add some number so wrong results are returned to see whether the unit tests actually cover some part and raise a failure correctly.
You could try with munp to see if there are unit tests for it.

seanpquinn commented:
For some reason a Python3-32 `scipy.signal` test now fails in appveyor? Unsure how this can be related to last commit involving moments.

ev-br commented:
LGTM too, merging. Thanks @seanpquinn, all.
Congratulations with your first scipy PR, and keep them coming!

There's a small glitch with the math in the docstring, will send a fix in a follow-up PR.
https://5136-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.stats.moyal.html#scipy.stats.moyal

ev-br commented:
And of course, a tutorial entry would be welcome :-).
